File: apps/app-house-keeper/android/arsenal/libraries/library-base/src/main/java/com/smart/library/util/hybird/HKHybirdManager.kt
package com.smart.library.util.hybird

import android.net.Uri
import android.text.TextUtils
import android.webkit.WebResourceResponse
import android.webkit.WebView
import com.smart.library.util.HKLogUtil
import com.smart.library.util.HKReflectUtil
import java.lang.reflect.InvocationTargetException
import java.net.URLDecoder
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ConcurrentMap
import kotlin.reflect.KClass
import kotlin.reflect.KFunction
import kotlin.reflect.full.companionObject
import kotlin.reflect.full.declaredFunctions

@Suppress("MemberVisibilityCanPrivate", "unused")
object HKHybirdManager {
    val TAG = "[hybird]"

    private val classMap: HashMap<String, KClass<*>> = hashMapOf()
    private val schemeMap: HashMap<String, (webView: WebView?, url: Uri?) -> Boolean> = hashMapOf()
    private val requestMap: HashMap<String, (webView: WebView?, url: String?) -> WebResourceResponse?> = hashMapOf()

    @Throws(RuntimeException::class, IllegalAccessException::class, IllegalArgumentException::class, InvocationTargetException::class, NullPointerException::class, ExceptionInInitializerError::class)
    fun callNativeMethod(className: String?, methodName: String?, vararg params: Any?): Any? {
        if (!classMap.containsKey(className)) throw RuntimeException("[callNativeMethod] class not register :$className")
        return HKReflectUtil.invoke(classMap[className], methodName, params)
    }

    /**
     * hybird://native/className/methodName?params=1,2,3,4,5&hashcode=123445
     */
    @Throws(RuntimeException::class, IllegalAccessException::class, IllegalArgumentException::class, InvocationTargetException::class, NullPointerException::class, ExceptionInInitializerError::class)
    fun addNativeClass(scheme: String, className: String, kClass: KClass<*>?) {
        if (TextUtils.isEmpty(className) || kClass == null) {
            throw RuntimeException("[addNativeClass] className:$className or kClass:$kClass is null")
        }
        if (!classMap.containsKey(className)) {
            val methods = HashMap<String, KFunction<*>>()
            kClass.java.kotlin.companionObject?.declaredFunctions?.forEach { methods.put(it.name, it) }
            classMap.put(className, kClass)
        }

        schemeMap.put(scheme, { webView: WebView?, schemeUrl: Uri? ->
            val pathSegments = schemeUrl?.pathSegments ?: arrayListOf()
            if (pathSegments.size >= 2) {
                val clazzName = pathSegments[0]
                val methodName = pathSegments[1]

                val queryMap: HashMap<String?, String?> = HashMap()
                schemeUrl?.encodedQuery?.split("&")?.forEach { it.split("=").takeIf { it.isNotEmpty() }?.apply { queryMap.put(this.getOrNull(0), this.getOrNull(1)) } }
                val hashcode = queryMap["hashcode"]  // val hashcode = url?.getQueryParameter("hashcode") // getQueryParameter 默认 decode
                val params = queryMap["params"]

                val paramArray = params?.split(",")?.map { URLDecoder.decode(it, "UTF-8") }?.toTypedArray()

                HKLogUtil.d(TAG, "clazzName:$clazzName , methodName:$methodName , params:size:${paramArray?.size}:($params) , hashCode:$hashcode")
                HKLogUtil.w(TAG, "native.invoke start --> [${kClass.java.name}.$methodName($params)]")

                var result: Any? = null
                try {
                    result = callNativeMethod(className, methodName, *paramArray ?: arrayOf())
                } catch (e: Exception) {
                    HKLogUtil.e(TAG, "native.invoke exception", e)
                }

                HKLogUtil.w(TAG, "native.invoke end , result:$result")
                if (!TextUtils.isEmpty(hashcode)) {
                    callJsFunction(webView, "javascript:window.hybird.onCallback($hashcode, $result)")
                }
                true
            } else {
                HKLogUtil.e(TAG, "schemaUrl:${schemeUrl.toString()} 格式定义错误，请参照 hybird://native/className/methodName?params=1,2,3,4,5&hashcode=123445")
                false
            }
        })
    }

    fun addRequest(host: String, intercept: (webView: WebView?, url: String?) -> WebResourceResponse?) {
        requestMap.put(host, intercept)
    }

    fun removeRequest(host: String) {
        requestMap.remove(host)
    }

    fun shouldInterceptRequest(webView: WebView?, url: String?): WebResourceResponse? {
        return requestMap.entries.filter { entry ->
            url?.contains(entry.key, true) ?: false
        }.getOrNull(0)?.value?.invoke(webView, url)


        //return requestMap[Uri.parse(url).host]?.invoke(webView, url)
    }

    /**
     *  scheme://host:port/path?k=v
     *
     *  schema  :   http            Gets the scheme of this URI. Example: "http"
     *  host    :   baidu.com       if the authority is "bob@google.com", this method will return "google.com".
     *  port    :   7777            if the authority is "google.com:80", this method will return 80.
     *  path    :   index.html
     */
    fun addScheme(scheme: String, host: String, port: Int, intercept: (webView: WebView?, url: Uri?) -> Boolean) {
        val schemePrefix = "$scheme://$host:$port"
        schemeMap.put(schemePrefix, intercept)
    }

    /**
     *  scheme://host:port/path?k=v
     *
     *  schema  :   http            Gets the scheme of this URI. Example: "http"
     *  host    :   baidu.com       if the authority is "bob@google.com", this method will return "google.com".
     *  port    :   7777            if the authority is "google.com:80", this method will return 80.
     *  path    :   index.html
     */
    fun addScheme(schemeUrlString: String, intercept: (webView: WebView?, url: Uri?) -> Boolean) {
        schemeMap.put(schemeUrlString, intercept)
    }

    fun removeScheme(schemeUrlString: String) {
        schemeMap.remove(schemeUrlString)
    }

    fun removeScheme(scheme: String, host: String, port: Int) {
        val schemePrefix = "$scheme://$host:$port"
        schemeMap.remove(schemePrefix)
    }

    fun shouldOverrideUrlLoading(webView: WebView?, uriString: String?): Boolean? {
        val url = Uri.parse(uriString)
        val schemePrefix = "${url?.scheme}://${url?.host}:${url?.port}"
        HKLogUtil.v(TAG, "[intercept?${schemeMap.containsKey(schemePrefix)}] , url=$url , schemePrefix=$schemePrefix")
        return schemeMap[schemePrefix]?.invoke(webView, url)
    }

    //========================================================================================================================
    // async callback
    //========================================================================================================================

    private val callbackMap: ConcurrentMap<String, ((result: String?) -> Unit?)?> = ConcurrentHashMap()
    private val callbackSchemaPrefix: String = "hybird://hybird:${(-System.currentTimeMillis().toInt())}"
    fun addSchemeForCallback() {
        addScheme(callbackSchemaPrefix) { _: WebView?, url: Uri? ->
            val hashCode = url?.getQueryParameter("hashcode")
            callbackMap[hashCode]?.invoke(url?.getQueryParameter("result"))
            callbackMap.remove(hashCode)
            HKLogUtil.e(HKHybirdManager.TAG, "callbackMap.size[after callback]:" + callbackMap.size)
            true
        }
    }

    fun callJsFunction(webView: WebView?, javascript: String, callback: ((result: String?) -> Unit?)? = null) {
        if (!schemeMap.containsKey(callbackSchemaPrefix))
            addSchemeForCallback()

        val jsString = javascript.replace("javascript:", "")
        val callbackHashCode: String? = callback?.hashCode()?.toString() + System.currentTimeMillis()
        if (callback != null)
            callbackMap.put(callbackHashCode, callback)

        HKLogUtil.e(HKHybirdManager.TAG, "callbackMap.size[before callback]:" + callbackMap.size + " :$javascript")

        val wrappedJavascript = """
                (function (){
                    console.log(" [wrap(________)] 　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　");
                    console.log(" [wrap(start___)] >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>");

                    var result = '';

                    try {
                        result=$jsString;
                        console.log(" [wrap(execute_)] result= "+result);
                    } catch(error) {
                        result = 'error';
                        console.log(" [wrap(execute_)] result= "+result+" error= "+error);
                    }

                    var jumpUrl = "$callbackSchemaPrefix/callback?hashcode=$callbackHashCode&result="+result;
                    console.log("\n[wrap(callback)] jumpUrl=" + jumpUrl);

                    (function (_jumpUrl) {
                        var c = document.createElement("div");
                        c.innerHTML = '<iframe style="display: none;" src="' + _jumpUrl + '"/>';
                        document.querySelector("body").appendChild(c);
                        setTimeout(function () {
                            document.querySelector("body").removeChild(c);
                        }, 1000);
                    })(jumpUrl);

                    console.log(" [wrap(end_____)] <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<< \n");
                })();
            """.trimIndent()

        webView?.loadUrl("javascript:$wrappedJavascript")
    }

    init {
        addNativeClass("hybird://hybird:1234", "native", HKHybirdMethods::class)
    }
}

File: apps/app-house-keeper/ios/arsenal/libraries/library-base/src/util/hybird/HKHybirdManager.swift
//
// Created by maokangren on 2017/10/26.
// Copyright (c) 2017 com.xixi. All rights reserved.
//

import Foundation
import WebKit
import UIKit

class HKHybirdManager {
    private static let TAG = "hybird"

    private static var classMap = [String: String]()
    private static var schemeMap = [String: (webView: WKWebView?, url: URL?) -> Bool?]()

    static func callNativeMethod(className: String, methodName: String, _ params: Any...) -> Any? {
        return HKReflectUtil.invoke(className, methodName, params)
    }

    /**
     * hybird://native/className/methodName?params=1,2,3,4,5&hashcode=123445
     */
    static func addNativeClass(scheme: String, virtualClassName: String, _ realClassName: String) {
        if (TextUtils.isEmpty(virtualClassName) || TextUtils.isEmpty(realClassName)) {
            HKLogUtil.d(TAG, "[addNativeClass] className:$className or kClass:$kClass is null")
            return
        }
        classMap[virtualClassName] = realClassName

        schemeMap[scheme] = { webView, schemeUrl in
            HKLogUtil.d(TAG, "schemeUrl:", schemeUrl?.absoluteString ?? "nil")
            let pathSegments = schemeUrl?.pathComponents ?? []
            HKLogUtil.d(TAG, "scheme:", schemeUrl?.scheme ?? "nil")
            HKLogUtil.d(TAG, "host:", schemeUrl?.host ?? "nil")
            HKLogUtil.d(TAG, "port:", schemeUrl?.port ?? "nil")
            HKLogUtil.d(TAG, "path:", schemeUrl?.path ?? "nil")
            HKLogUtil.d(TAG, "lastPathComponent:", schemeUrl?.lastPathComponent ?? "nil")
            HKLogUtil.d(TAG, "pathSegments:", pathSegments)
            HKLogUtil.d(TAG, "query:", schemeUrl?.query ?? "nil")
            HKLogUtil.d(TAG, "queryParams:", schemeUrl?.queryParams ?? "nil")
            HKLogUtil.d(TAG, "query:key:params:", schemeUrl?.query("params") ?? "nil")
            HKLogUtil.d(TAG, "queryParamsUnDecoded:", schemeUrl?.queryParamsUnDecoded ?? "nil")
            HKLogUtil.d(TAG, "queryUnDecoded:key:params:", schemeUrl?.queryUnDecoded("params") ?? "nil")

            if (pathSegments.count >= 3) {
                let clazzName = classMap[pathSegments[1]] ?? pathSegments[1]
                let methodName = pathSegments[2]

                let hashcode = schemeUrl?.queryUnDecoded("hashcode") ?? ""
                let params = schemeUrl?.queryUnDecoded("params")

                let paramArray: [String] = params?.split(separator: ",").map {
                    $0.string?.urlDecoded ?? ""
                } ?? []

                HKLogUtil.d(TAG, "clazzName:\(clazzName) , methodName:\(methodName) , params:size:\(String(describing: paramArray.count)):(\(String(describing: paramArray)) , hashCode:\(String(describing: hashcode))")
                HKLogUtil.w(TAG, "native.invoke start --> [\(clazzName).\(methodName)(\(paramArray)]")
//
                var result: Any? = nil
                do {
                    try result = callNativeMethod(className: clazzName, methodName: methodName, paramArray)
                } catch {
                    HKLogUtil.e(TAG, "native.invoke exception", error)
                }
//
                HKLogUtil.w(TAG, "native.invoke end , result:\(String(describing: result))")
//                if (!TextUtils.isEmpty(hashcode)) {
//                    callJsFunction(webView, "javascript:window.hybird.onCallback($hashcode, $result)")
//                }
                return true
            } else {
                HKLogUtil.e(TAG, "schemaUrl:${schemeUrl.toString()} 格式定义错误，请参照 hybird://native/className/methodName?params=1,2,3,4,5&hashcode=123445")
                return false
            }
        }
    }

    /**
     *  scheme://host:port/path?k=v
     *
     *  schema  :   http            Gets the scheme of this URI. Example: "http"
     *  host    :   baidu.com       if the authority is "bob@google.com", this method will return "google.com".
     *  port    :   7777            if the authority is "google.com:80", this method will return 80.
     *  path    :   index.html
     */
    public static func addScheme(_ schemeUrlString: String, _ intercept: @escaping (_ webView: WKWebView?, _ url: URL?) -> Bool) {
        schemeMap[schemeUrlString] = intercept
    }

    public static func removeScheme(schemeUrlString: String) {
        schemeMap.removeValue(forKey: schemeUrlString)
    }

    public static func removeScheme(scheme: String, host: String, port: Int) {
        let schemePrefix = "\(scheme)://\(host):\(port)"
        schemeMap.removeValue(forKey: schemePrefix)
    }

    public static func shouldOverrideUrlLoading(_ webView: WKWebView?, _ userContentController: WKUserContentController, _ message: WKScriptMessage) -> Bool? {
        let schemeUrl = URL(string: String(describing: message.body))
        let schemePrefix = "\(String(describing: schemeUrl?.scheme ?? ""))://\(String(describing: schemeUrl?.host ?? "")):\(String(describing: schemeUrl?.port ?? 0))"
        HKLogUtil.d(HKHybirdManager.TAG, "intercept message.body : \(String(describing: message.body))")
        HKLogUtil.d(HKHybirdManager.TAG, "intercept url : \(String(describing: schemeUrl?.absoluteString ?? ""))")
        HKLogUtil.d(HKHybirdManager.TAG, "intercept schemePrefix : \(schemePrefix)")
        HKLogUtil.d(HKHybirdManager.TAG, "intercept ? \(schemeMap.containsKey(schemePrefix))")
        return schemeMap[schemePrefix]?(webView, schemeUrl)
    }

    public static func callJsFunction(_ webView: WKWebView?, _ javascript: String, _ callback: ((_ result: String?) -> Void?)? = nil) {
        webView?.evaluateJavaScript(javascript) { (any: Any?, error: Error?) in
            HKLogUtil.d(TAG, "any", (any ?? "nil"), "error", (error ?? "nil"))
            let result = error == nil ? (any as? String) : ("error:" + error.debugDescription)
            callback?(result)
        }
    }
}
